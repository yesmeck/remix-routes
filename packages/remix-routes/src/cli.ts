#!/usr/bin/env node
import meow from 'meow';
import * as fs from 'fs';
import * as path from 'path';
import chokidar from 'chokidar';
import type { ConfigRoute } from '@remix-run/dev/dist/config/routes';
import mkdirp from 'mkdirp';

let readConfig: typeof import('@remix-run/dev/dist/config').readConfig;

try {
  readConfig = require('@remix-run/dev/config').readConfig;
} catch (e) {
  readConfig = require('@remix-run/dev/dist/config').readConfig;
}

const helpText = `
Usage
$ remix-routes

Options
--watch, -w  Watch for routes changes
--outputDirPath, -o Specify the output path for "remix-routes.d.ts". Defaults to "./node_modules" if arg is not given.
`;

const DEFAULT_OUTPUT_DIR_PATH = './node_modules'

const cli = meow(helpText, {
  flags: {
    watch: {
      type: 'boolean',
      alias: 'w',
    },
    outputDirPath: {
      type: 'string',
      alias: 'o',
      default: DEFAULT_OUTPUT_DIR_PATH,
    }
  },
});

type RoutesInfo = Record<string, {
  fileName: string;
  params: string[];
}>

async function buildHelpers(remixRoot: string): Promise<RoutesInfo> {
  const config = await readConfig(remixRoot);
  const routesInfo: RoutesInfo = {};
  const handleRoutesRecursive = (
    parentId?: string,
    parentPath: ConfigRoute[] = [],
  ) => {
    let routes = Object.values(config.routes).filter(
      (route) => route.parentId === parentId,
    );
    routes.forEach((route) => {
      let currentPath = parentPath;
      if (route.path) {
        currentPath = [...currentPath, route];
        const fullPath = currentPath.reduce(
          (acc, curr) => [acc, trimSlash(curr.path!)].join('/'),
          '',
        );
        const paramsNames = parse(currentPath);
        routesInfo[fullPath] = {
          fileName: route.file,
          params: paramsNames
        };
      } else if (route.id === 'root') {
        routesInfo['/'] = {
          fileName: route.file,
          params: [],
        };
      }
      handleRoutesRecursive(route.id, currentPath);
    });
  };
  handleRoutesRecursive();
  return routesInfo;
}

export async function build(remixRoot: string, outputDirPath: string) {
  const routesInfo = await buildHelpers(remixRoot);
  generate(routesInfo, remixRoot, outputDirPath);
}

function watch(remixRoot: string, outputDirPath: string) {
  build(remixRoot, outputDirPath);
  chokidar
    .watch([
      path.join(remixRoot, 'app/routes/**/*.{ts,tsx}'),
      path.join(remixRoot, 'remix.config.js'),
    ])
    .on('change', () => {
      build(remixRoot, outputDirPath);
    });
  console.log('Watching for routes changes...');
}

function generate(routesInfo: RoutesInfo, remixRoot: string, outputDirPath: string) {
  const tsCode =
    [
      `
declare module "remix-routes" {

type IsAny<T> = (
  unknown extends T
    ? [keyof T] extends [never] ? false : true
    : false
);
type URLSearchParamsInit = string | string[][] | Record<string, string> | URLSearchParams;
type Query<T> = IsAny<T> extends true ? URLSearchParamsInit : T;
      `,
      generateRouteDefinition(routesInfo),
      generatePathDefinition(routesInfo),
      generateParamsDefinition(routesInfo),
      "}",
    ].join('\n\n') + '\n\n';

  const outputPath = path.join(
    remixRoot,
    outputDirPath,
  );

  if (!fs.existsSync(outputPath)) {
    mkdirp.sync(outputPath);
  }
  fs.writeFileSync(path.join(outputPath, 'remix-routes.d.ts'), tsCode);
}

function generateRouteDefinition(routesInfo: RoutesInfo) {
  const code: string[] = ['export interface Routes {'];
  Object.entries(routesInfo).forEach(([route, { fileName, params }]) => {
    const lines = [`  "${route}": {`];
    const paramsType = params.map(
      (param) => `${param}: string | number`,
    );
    lines.push(`    params: { ${paramsType.join('; ')} },`);
    lines.push(`    query: Query<import('../app/${fileName.replace(/\.tsx?$/, '')}').SearchParams>,`)
    lines.push('  };')
    code.push(lines.join('\n'));
  });
  code.push("}");
  return code.join('\n');
}

function generatePathDefinition(routesInfo: RoutesInfo) {
  const code: string[] = [];
  Object.entries(routesInfo).forEach(([route, { fileName, params }]) => {
    const lines = ['export declare function $path('];
    lines.push(`  route: ${JSON.stringify(route)},`);
    if (params.length > 0) {
      lines.push(`  params: Routes["${route}"]["params"],`);
    }
    lines.push(
      `  ...query: [Routes["${route}"]["query"]]`,
    );
    lines.push(`): string;`);
    code.push(lines.join('\n'));
  });
  return code.join('\n');
}

function generateParamsDefinition(routesInfo: RoutesInfo) {
  const routes = Object.entries(routesInfo);

  // $params helper makes sense only for routes with params.
  const routesWithParams = routes.filter(
    ([_, { params }]) => params.length > 0,
  );

  const code = routesWithParams.map(([route, { params }]) => {
    const lines: string[] = [];

    lines.push(`export declare function $params(`);
    lines.push(`  route: ${JSON.stringify(route)},`);
    lines.push(`  params: { readonly [key: string]: string | undefined }`);
    lines.push(`): {`);
    lines.push(
      params.map((param) => `  ${param}: string`).join(',\n'),
    );
    lines.push(`};`);

    return lines.join('\n');
  });

  return code.join('\n');
}

function parse(routes: ConfigRoute[]) {
  const paramNames: string[] = [];
  routes.forEach((route) => {
    return (
      route.path &&
      paramNames.push(
        ...route.path
          .split('/')
          .filter((seg) => seg.startsWith(':'))
          .map((param) => param.replace(':', '')),
      )
    );
  });
  return paramNames;
}

if (require.main === module) {
  (async function () {
    const remixRoot = process.env.REMIX_ROOT || process.cwd();

    if (cli.flags.watch) {
      watch(remixRoot, cli.flags.outputDirPath);
    } else {
      build(remixRoot, cli.flags.outputDirPath);
    }
  })();
}

function trimSlash(path: string) {
  return path.replace(/\/+$/, '');
}
