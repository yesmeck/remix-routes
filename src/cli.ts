#!/usr/bin/env node
import meow from 'meow';
import * as fs from 'fs';
import * as path from 'path';
import chokidar from 'chokidar';
import type { ConfigRoute } from '@remix-run/dev/dist/config/routes';
import mkdirp from 'mkdirp';

let readConfig: typeof import('@remix-run/dev/dist/config').readConfig;

try {
  readConfig = require('@remix-run/dev/config').readConfig;
} catch (e) {
  readConfig = require('@remix-run/dev/dist/config').readConfig;
}

const helpText = `
Usage
$ remix-routes

Options
--watch, -w  Watch for routes changes
`;

const cli = meow(helpText, {
  flags: {
    watch: {
      type: 'boolean',
      alias: 'w',
    },
  },
});

type RoutesInfo = Record<string, string[]>;

async function buildHelpers(remixRoot: string): Promise<RoutesInfo> {
  const config = await readConfig(remixRoot);
  const routesInfo: RoutesInfo = {};
  const handleRoutesRecursive = (
    parentId?: string,
    parentPath: ConfigRoute[] = [],
  ) => {
    let routes = Object.values(config.routes).filter(
      (route) => route.parentId === parentId,
    );
    routes.forEach((route) => {
      let currentPath = parentPath;
      if (route.path) {
        currentPath = [...currentPath, route];
        const fullPath = currentPath.reduce(
          (acc, curr) => [acc, trimSlash(curr.path!)].join('/'),
          '',
        );
        const paramsNames = parse(currentPath);
        routesInfo[fullPath] = paramsNames;
      }
      handleRoutesRecursive(route.id, currentPath);
    });
  };
  handleRoutesRecursive();
  return routesInfo;
}

export async function build(remixRoot: string) {
  const routesInfo = await buildHelpers(remixRoot);
  generate(routesInfo);
}

function watch(remixRoot: string) {
  build(remixRoot);
  chokidar
    .watch([
      path.join(remixRoot, 'app/routes/**/*.{ts,tsx}'),
      path.join(remixRoot, 'remix.config.js'),
    ])
    .on('change', () => {
      build(remixRoot);
    });
  console.log('Watching for routes changes...');
}

function generate(routesInfo: RoutesInfo) {
  const jsCode = generateHelpers(routesInfo);
  const tsCode =
    [
      generatePathDefinition(routesInfo),
      generateParamsDefinition(routesInfo),
    ].join('\n\n') + '\n\n';

  const outputPath = path.join(
    process.cwd(),
    'node_modules',
    'remix-routes',
  );

  if (!fs.existsSync(outputPath)) {
    mkdirp.sync(outputPath);
  }
  fs.writeFileSync(path.join(outputPath, 'index.js'), jsCode);
  fs.writeFileSync(path.join(outputPath, 'index.d.ts'), tsCode);
  fs.writeFileSync(
    path.join(outputPath, 'package.json'),
    JSON.stringify({
      name: 'remix-routes',
      main: 'index.js',
    }),
  );
}

function generateHelpers(routesInfo: RoutesInfo) {
  const routes = Object.entries(routesInfo).reduce(
    (acc: Record<string, string[]>, [route, params]) => {
      if (params.length > 0) {
        acc[route] = params;
      }
      return acc;
    },
    {},
  );
  const helper = fs.readFileSync(path.resolve(__dirname, '../helper.js'), 'utf-8')
  return helper.replace('// __routes__', `const routes = ${JSON.stringify(routes, null, 2)}`);
}

function generatePathDefinition(routesInfo: RoutesInfo) {
  const code: string[] = [];
  Object.entries({
    '/': [],
    ...routesInfo,
  }).forEach(([route, paramsNames]) => {
    const lines = ['export declare function $path('];
    lines.push(`  route: ${JSON.stringify(route)},`);
    if (paramsNames.length > 0) {
      const paramsType = paramsNames.map(
        (paramName) => `${paramName}: string | number`,
      );
      lines.push(`  params: { ${paramsType.join('; ')} },`);
    }
    lines.push(
      `  query?: string | string[][] | Record<string, string> | URLSearchParams`,
    );
    lines.push(`): string;`);
    code.push(lines.join('\n'));
  });
  return code.join('\n');
}

function generateParamsDefinition(routesInfo: RoutesInfo) {
  const routes = Object.entries(routesInfo);

  // $params helper makes sense only for routes with params.
  const routesWithParams = routes.filter(
    ([_, paramsNames]) => paramsNames.length > 0,
  );

  const code = routesWithParams.map(([route, paramsNames]) => {
    const lines: string[] = [];

    lines.push(`export declare function $params(`);
    lines.push(`  route: ${JSON.stringify(route)},`);
    lines.push(`  params: { readonly [key: string]: string | undefined }`);
    lines.push(`): {`);
    lines.push(
      paramsNames.map((paramName) => `  ${paramName}: string`).join(',\n'),
    );
    lines.push(`};`);

    return lines.join('\n');
  });

  return code.join('\n');
}

function parse(routes: ConfigRoute[]) {
  const paramNames: string[] = [];
  routes.forEach((route) => {
    return (
      route.path &&
      paramNames.push(
        ...route.path
          .split('/')
          .filter((seg) => seg.startsWith(':'))
          .map((param) => param.replace(':', '')),
      )
    );
  });
  return paramNames;
}

if (require.main === module) {
  (async function () {
    const remixRoot = process.env.REMIX_ROOT || process.cwd();

    if (cli.flags.watch) {
      watch(remixRoot);
    } else {
      build(remixRoot);
    }
  })();
}

function trimSlash(path: string) {
  return path.replace(/\/+$/, '');
}
