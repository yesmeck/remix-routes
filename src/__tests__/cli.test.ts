import { build } from '../cli';
import * as path from 'path';
import * as fs from 'fs';

test('build', async () => {
  await build(path.resolve(__dirname, '../../fixture'));
  const outputPath = path.resolve(__dirname, '../../node_modules');
  fs.readdirSync(outputPath).forEach((file) => {
    if (file.startsWith('remix-routes')) {
      expect(
        fs.readFileSync(path.resolve(outputPath, file), 'utf8'),
      ).toMatchSnapshot();
    }
  });
});
