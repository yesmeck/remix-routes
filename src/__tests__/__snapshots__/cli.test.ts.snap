// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`build 1`] = `
"
const routes = {
  \\"/chats/:season/:episode/:slug\\": [
    \\"season\\",
    \\"episode\\",
    \\"slug\\"
  ],
  \\"/chats/:season/:episode\\": [
    \\"season\\",
    \\"episode\\"
  ],
  \\"/people/:personId\\": [
    \\"personId\\"
  ],
  \\"/people/:personId/:planId/remove-plan\\": [
    \\"personId\\",
    \\"planId\\"
  ],
  \\"/s/:query\\": [
    \\"query\\"
  ],
  \\"/admin/episodes/:id\\": [
    \\"id\\"
  ],
  \\"/admin/episodes/:id/comments\\": [
    \\"id\\"
  ],
  \\"/jokes/:jokeId\\": [
    \\"jokeId\\"
  ]
};

module.exports = { routes }
"
`;

exports[`build 2`] = `"{\\"name\\":\\".remix-routes\\",\\"main\\":\\"index.js\\"}"`;

exports[`build 3`] = `
"export declare function $path(
  route: \\"/\\",
  query?: Record<string, string | number>
): string;
export declare function $path(
  route: \\"/chats/:season/:episode/:slug\\",
  params: { season: string | number; episode: string | number; slug: string | number },
  query?: Record<string, string | number>
): string;
export declare function $path(
  route: \\"/chats/:season/:episode\\",
  params: { season: string | number; episode: string | number },
  query?: Record<string, string | number>
): string;
export declare function $path(
  route: \\"/people/:personId\\",
  params: { personId: string | number },
  query?: Record<string, string | number>
): string;
export declare function $path(
  route: \\"/people/:personId/:planId/remove-plan\\",
  params: { personId: string | number; planId: string | number },
  query?: Record<string, string | number>
): string;
export declare function $path(
  route: \\"/blog/rss.xml\\",
  query?: Record<string, string | number>
): string;
export declare function $path(
  route: \\"/auth\\",
  query?: Record<string, string | number>
): string;
export declare function $path(
  route: \\"/auth/login\\",
  query?: Record<string, string | number>
): string;
export declare function $path(
  route: \\"/s/:query\\",
  params: { query: string | number },
  query?: Record<string, string | number>
): string;
export declare function $path(
  route: \\"/credits\\",
  query?: Record<string, string | number>
): string;
export declare function $path(
  route: \\"/admin\\",
  query?: Record<string, string | number>
): string;
export declare function $path(
  route: \\"/admin/episodes\\",
  query?: Record<string, string | number>
): string;
export declare function $path(
  route: \\"/admin/episodes/:id\\",
  params: { id: string | number },
  query?: Record<string, string | number>
): string;
export declare function $path(
  route: \\"/admin/episodes/:id/comments\\",
  params: { id: string | number },
  query?: Record<string, string | number>
): string;
export declare function $path(
  route: \\"/admin/episodes/new\\",
  query?: Record<string, string | number>
): string;
export declare function $path(
  route: \\"/jokes\\",
  query?: Record<string, string | number>
): string;
export declare function $path(
  route: \\"/jokes/:jokeId\\",
  params: { jokeId: string | number },
  query?: Record<string, string | number>
): string;

export declare function $params(
  route: \\"/chats/:season/:episode/:slug\\",
  params: { readonly [key: string]: string | undefined }
): {
  season: string,
  episode: string,
  slug: string
};
export declare function $params(
  route: \\"/chats/:season/:episode\\",
  params: { readonly [key: string]: string | undefined }
): {
  season: string,
  episode: string
};
export declare function $params(
  route: \\"/people/:personId\\",
  params: { readonly [key: string]: string | undefined }
): {
  personId: string
};
export declare function $params(
  route: \\"/people/:personId/:planId/remove-plan\\",
  params: { readonly [key: string]: string | undefined }
): {
  personId: string,
  planId: string
};
export declare function $params(
  route: \\"/s/:query\\",
  params: { readonly [key: string]: string | undefined }
): {
  query: string
};
export declare function $params(
  route: \\"/admin/episodes/:id\\",
  params: { readonly [key: string]: string | undefined }
): {
  id: string
};
export declare function $params(
  route: \\"/admin/episodes/:id/comments\\",
  params: { readonly [key: string]: string | undefined }
): {
  id: string
};
export declare function $params(
  route: \\"/jokes/:jokeId\\",
  params: { readonly [key: string]: string | undefined }
): {
  jokeId: string
};

"
`;
